brainframe_qt/ui/resources/ui_elements/applications/messaging_application.py

```diff
@@ -0,0 +1,364 @@
+
+        if socket.canReadLine():
+            # Immediate ready to read data
+            logging.debug("immediate _handle_ready_read")
+            self._handle_ready_read()
+        else:
+            logging.debug("connect _handle_ready_read")
+            socket.readyRead.connect(self._handle_ready_read)
+
+        if socket.error() not in [
+            socket.UnknownSocketError, socket.PeerClosedError
+        ]:
+            # Immediate error
+            logging.debug("immediate _handle_socket_error")
+            self._handle_socket_error(socket.error())
+        else:
+            logging.debug("connect _handle_socket_error")
+            socket.error.connect(self._handle_socket_error)
+
+        if socket.state() == QLocalSocket.UnconnectedState:
+            # Immediate disconnect
+            logging.debug("immediate _handle_disconnect")
+            self._handle_disconnect()
+        else:
+            logging.debug("connect _handle_disconnect")
+            socket.disconnected.connect(self._handle_disconnect)
+
+    def _handle_disconnect(self) -> None:
+        logging.debug("exec _handle_disconnect")
+
+        if self.previous_connection is not None:
+            self.previous_connection.deleteLater()
+
+        socket = self.current_connection
+        logging.debug(f"Client disconnected from server: 0x{id(socket):x}")
+
+        self.previous_connection, self.current_connection = (
+            self.current_connection, None
+        )
+
+        # Deal with potential queue of connections
+        self._handle_connection()
+
+    def _handle_ready_read(self) -> None:
+        logging.debug("exec _handle_ready_read")
+
+        socket = self._get_socket()
+        if socket is None:
+            logging.error(
+                "Attempted to read message from socket, but socket has already been "
+                "disconnected/deleted"
+            )
+            return
+
+        message_data = bytes(socket.readAll())
+        logging.debug(f"Received raw data from socket 0x{id(socket):x}: {message_data}")
+
+        message = self._parse_message_data(message_data)
+        self.new_message.emit(message)
+
+    def _handle_socket_error(self, error: QLocalSocket.LocalSocketError) -> None:
+        logging.debug(f"exec _handle_socket_error | {error}")
+        socket = self._get_socket()
+
+        if socket is None:
+            logging.error(
+                "Attempted to deal with socket error, but socket has already been "
+                "disconnected/deleted"
+            )
+            return
+
+        if socket.error() == socket.PeerClosedError:
+            logging.debug(f"Peer disconnected on socket 0x{id(socket):x}")
+        else:
+            error_message = (
+                f"Error occurred with IPC socket 0x{id(socket):x}: "
+                f"{error} | {self.errorString()}"
+            )
+            logging.error(error_message)
+            raise self.UnknownSocketError(error_message)
```

Comment:
**Exception**: 
```
Traceback (most recent call last):
  File "brainframe_qt/ui/resources/ui_elements/applications/messaging_application.py", line 270, in _handle_socket_error
brainframe_qt.ui.resources.ui_elements.applications.messaging_application.MessagingServer.UnknownSocketError: Error occurred with IPC socket 0x7f361093d678: 1 | 
```

**Problem**: 
When a client is already open, sometimes this error happens to the client that is supposed to give way to the first client. 

**Suggested resolution**: 
I feel like unknown errors shouldn't raise, but maybe log and close the client. Otherwise users just get these giant error messages that proceed to... close the client anyways.

I don't think this is that big of a deal, but I do think maybe we shouldn't be popping windows open for this. I'm not sure. 

**Replicate**:
```
bash brainframe_client.sh & bash brainframe_client.sh & bash brainframe_client.sh
```

This doesn't seem to happen with the snap, because the snap has a slower startup time and that prevents the race condition from happening.

Reply:
Ok it looks like this is what's happening:

1. Two (or more) clients start at ~roughly the same time. Both try to acquire the singleton lock:
   https://github.com/aotuai/brainframe-qt/blob/c116edd8ecea7b19d0d56878c9a6e7c7a3973109/brainframe_qt/ui/resources/ui_elements/applications/singleton_application.py#L22-L29
   
2. The first instance initializes as a server
3. At the same time the second instance begins to initialize as a client

4. At the end of the initialization sequence for client instances, we send a message to the instance that acts as the server.
   https://github.com/aotuai/brainframe-qt/blob/c116edd8ecea7b19d0d56878c9a6e7c7a3973109/brainframe_qt/ui/resources/ui_elements/applications/singleton_application.py#L44-L45
   https://github.com/aotuai/brainframe-qt/blob/c116edd8ecea7b19d0d56878c9a6e7c7a3973109/brainframe_qt/ui/resources/ui_elements/applications/singleton_application.py#L55-L62
   
If the server instance has not gotten to a point where it's ready to accept messages, the client instance is not expecting this and triggers the exception that you're seeing.

---

I believe that I can fix this. I also think that you're point about just logging errors instead of raising real exceptions is a good one. I will make that change

Reply:
This has been addressed

Reply:
Did this for sure get resolved? I noticed UnknownSocketError still brings up a window. I was testing with the snap though, maybe I didn't have the right build?

Reply:
Do you have a stack trace for that? Potentially I'm not catching an exception properly

Reply:
Hmm... How do I get logs for each program, when they were run with `&`? Here's what the clients `Show Details` button gives me:

```
Traceback (most recent call last):
  File "brainframe_qt/ui/resources/ui_elements/applications/messaging_application.py", line 273, in _handle_socket_error
brainframe_qt.ui.resources.ui_elements.applications.messaging_application.MessagingServer.UnknownSocketError: Error occurred with IPC socket 0x7f6406966ca8: 1 | 
```

This is on a the `c772d77a` pyinstaller build.

Reply:
I have determined the issue. Fix incoming